llvm/lib/Target/WebAssembly/WebAssemblyAsmPrinter.cpp

```diff
@@ -609,6 +676,91 @@ void WebAssemblyAsmPrinter::emitJumpTableInfo() {
   // Nothing to do; jump tables are incorporated into the instruction stream.
 }
 
+std::optional<bool> WebAssemblyAsmPrinter::getBranchHint(const MachineInstr& MI) {
+  // WebAssemblyLowerBrUnless should have run before us, removing all BR_UNLESS,
+  // which makes things simpler for us here.
+  assert(MI.getOpcode() != WebAssembly::BR_UNLESS);
+
+  if (MI.getOpcode() != WebAssembly::BR_IF)
+    return {};
+
+  // We only handle the simple case of our being the last terminator of the
+  // block. If there are other terminators after us, things may be complicated.
+  auto *ParentMBB = MI.getParent();
+  const MachineInstr *LastTerminator = nullptr;
+  for (auto& Terminator : ParentMBB->terminators())
+    LastTerminator = &Terminator;
+  if (LastTerminator != &MI)
+    return {};
+
+  // This is a BR. It has two successors, and perhaps branch probability
+  // info between them. Finding the successor blocks is not trivial, since we
+  // run after CFGstackify (and even WebAssemblyInstrInfo::analyzeBranch returns
+  // that it cannot analyze branch targets). First, find the two successors of
+  // the parent block of this BR_IF. (If there are three successors, due to some
+  // terminator before us, give up on this hint.)
+  if (ParentMBB->succ_size() != 2)
+    return {};
+  auto iter = ParentMBB->succ_begin();
+  MachineBasicBlock* MBB1 = *iter;
+  iter++;
+  MachineBasicBlock* MBB2 = *iter;
+
+  // See which of the two successors is right after us: the fallthrough. If we
+  // branch, it is to the other block. So the block right after us is where we
+  // go if the br_if condition is false.
+  auto ParentMBBI = ParentMBB->getIterator();
+  ++ParentMBBI;
+  // A block with a BR_IF must have something after it.
+  assert(ParentMBBI != MF->end());
+  auto *FalseDest = &*ParentMBBI;
+  // It must be one of the successors.
+  assert(FalseDest == MBB1 || FalseDest == MBB2);
+  // Stop if the fallthrough is an EH pad, because that could happen like this:
+  //
+  //  bb1:
+  //    ;; successor: if.true, otherwise
+  //    ..
+  //    br_if $if.true
+  //    br $otherwise
+  //
+  //  ehpad:
+  //
+  // If that br $otherwise is optimized out, it would look like we fall through
+  // to the ehpad (since it is the block physically after us), but in wasm's
+  // structured control flow that is not the case. Rather than try to find the
+  // true fallthrough, give up on a hint in this rare case.
+  if (FalseDest->isEHPad())
+    return {};
```

Comment:
I think this can have another corner case too... For example
```
bb0:
  ...
  br_if

bb1:
  delegate

ehpad:
  catch
```

So this `delegate` here would have been inserted when we first place `try`-`catch` markers. We add `delegate` in a separate BB because it is also a terminator. In this case the fallthrough BB is not an EH pad but we still shouldn't fall through to the next BB.

I think a safer solution could be... Just bail out when one of the ParentBB's successor is not the next BB in the layout order. I think this can handle both cases.